### PR TITLE
DOC: NEP for C style guide

### LINKED ISSUE #11911: DOC: Make C_STYLE_GUIDE an NEP
### Reproducing code example:

```python
import numpy as np
<< your code here >>
```

### PR DESCRIPTION
**Pull request by bjnath**
Draft of NEP 45 to set out a coding style for NumPy C code. Recasts `doc/C_STYLE_GUIDE.rst.txt` in NEP style.

Fixes gh-11911 "DOC: Make C_STYLE_GUIDE an NEP"

### CODE REVIEW

**Comment by bashtage** [comment]
Will you produce a clang-format configuration?

**Comment by rgommers** [comment]
@bjnath it would be good to delete the old file. Or better, leave the file and replace the content with a link to the devdocs location that content moved to.

> Will you produce a clang-format configuration?

This is just moving a doc file into a better place, so that's a bit out of scope.

**Comment by bjnath** [comment]
I assume that file isn't visible through the website (I couldn't find it)?  I can just add a note to the top.

**Comment by rgommers** [comment]
> I assume that file isn't visible through the website (I couldn't find it)?

it's not indeed - it's just in the repo as a plain text file

**Comment by charris** [comment]
We will probably want to make some changes now that we require c99. The main changes I can see is `//` is now legal for comments and declaration of variables after a statement is now allowed. I still prefer declarations to be grouped at the top with the exception of loop variables, but that ship has sailed...

**Comment by bjnath** [comment]
Replaced contents of page with a pointer to NEP 45.

BTW, the HTML for the NEP (or for any NEP) doesn't appear after build.  Am I missing a step?

**Comment by rossbar** [comment]
> BTW, the HTML for the NEP (or for any NEP) doesn't appear after build. Am I missing a step?

This had me stumped too - the NEPs are built separate from the documentation with their own sphinx configuration. If you want to build the NEPs locally, you can do so via `cd doc/neps && make html`

File: doc/neps/nep-0045-c_style_guide.rst [hunk 1]
```diff
@@ -0,0 +1,269 @@
+=================================
+NEP 45 — C Style Guide
+=================================
+
+:Author: Charles Harris <charlesr.harris@gmail.com>
+:Author: Pierre de Buyl <pdebuyl@pdebuyl.be>
+:Author: Matti Picus <matti.picus@gmail.com>
+:Author: Warren Weckesser <warren.weckesser@gmail.com>
+:Author: Parul Aggarwal <parulagg27@gmail.com>
```
Comment (rgommers) [inline doc/neps/nep-0045-c_style_guide.rst:9]: @charris should be listed as the sole author here. He wrote the whole thing, all other commits to that file are minor tweaks.

File: doc/neps/nep-0045-c_style_guide.rst [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+:Status: Draft
```
Comment (rgommers) [inline doc/neps/nep-0045-c_style_guide.rst:10]: I think this should be Accepted. It's not a new proposal that should go through review. It's our existing process, changes to it are out of scope for this PR.

File: doc/neps/nep-0045-c_style_guide.rst [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+
+Related Work
+------------
+
+Based on Van Rossum and Warsaw, `PEP 7 -- Style Guide for C Code <https://www.python.org/dev/peps/pep-0007>`_
+
+
+
+Implementation
+--------------
```
Comment (rgommers) [inline doc/neps/nep-0045-c_style_guide.rst:255]: You can remove this section I think.

File: doc/neps/nep-0045-c_style_guide.rst [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+
+
+Discussion
+----------
+
+https://github.com/numpy/numpy/issues/11911
+recommended that this proposal, which originated as `doc/C_STYLE_GUIDE.rst.txt`,
```
Comment (rgommers) [inline doc/neps/nep-0045-c_style_guide.rst:262]: minor: double backticks needed around the filename here.

**Review by rgommers — COMMENTED** [review]
LGTM, only a couple of small comments. Thanks for doing this @bjnath

**Comment by rgommers** [comment]
> We will probably want to make some changes now that we require c99

I agree - but not in this PR.

**Comment by bjnath** [comment]
@rgommers 'Accepted' status requires a `Resolution` URL like `https://mail.python.org/pipermail/numpy-discussion/2020-March/080467.html` or else build fails with:
```
RuntimeError: NEP 45 is Accepted/Rejected/Withdrawn but has no Resolution tag
```
I can write one.

**Comment by rgommers** [comment]
Resolution can simply point to gh-11911 I think

**Comment by bjnath** [comment]
b6c917c  makes the requested revisions.

**Review by rgommers — APPROVED** [review]
LGTM now, thanks @bjnath

File: doc/neps/nep-0045-c_style_guide.rst [hunk 5]
```diff
@@ -0,0 +1,261 @@
+=================================
```
Comment (charris) [inline doc/neps/nep-0045-c_style_guide.rst:1]: Maybe put `.. _NEP45` up top to ease linking to this document. There is also a reference in `doc/source/dev/style_guide.rst` that needs a fix, not sure of the best way to do that.

Comment (charris) [reply doc/neps/nep-0045-c_style_guide.rst:1]: https://docutils.sourceforge.io/docs/ref/rst/directives.html#including-an-external-document-fragment explains the `include` directive, that may help.

Comment (rgommers) [reply doc/neps/nep-0045-c_style_guide.rst:1]: `doc/source/dev/style_guide.rst` can be removed, and `source/dev/index` can be changed to link to the NEP. 

I suggest to directly link to the absolute URL (`https://numpy.org/neps/nep-0045-...`), because the `_NEP45` anchor thing won't work well probably, given that the `doc/neps` is a completely separate Sphinx build and deploy.

Comment (charris) [reply doc/neps/nep-0045-c_style_guide.rst:1]: Do you know if the links are checked in the CircleCI builds?

Comment (rgommers) [reply doc/neps/nep-0045-c_style_guide.rst:1]: Not yet I believe, because we're not at zero warnings yet so can't turn on the check.

Easy to verify manually though via the ` ci/circleci: build artifact` link in the CI status list.

**Comment by mattip** [comment]
> I suggest to directly link to the absolute URL ...

We could use the intersphinx linking, `https://numpy.org/neps/objects.inv` exists.

File: doc/neps/nep-0045-c_style_guide.rst (hunk 5 above)
Comment (bjnath) [reply doc/neps/nep-0045-c_style_guide.rst:1]: I ran a checker when preparing my Season of Docs proposal. Nearly every page has a bad link because of the 'Created using Sphinx 2.4.4' line at the bottom. It links to `http://sphinx.pocoo.org/`, which returns a 400.

I found about two dozen other fails. Starting a PR.

**Comment by mattip** [comment]
Thanks @bjnath. Possible future cleanups:
- replace all the direct `https://numpy.org/neps` references by intersphinx linking (easily found via `git grep "numpy.org/neps")
- replace the `.. include` in `doc/source/dev/style_guide.rst` with a direct intersphinx link to the NEP.

**Comment by mattip** [comment]
Whoops: I didn't check the [newly rendered page](https://numpy.org/devdocs/dev/style_guide.html#style-guide) which now is empty. This is referenced from a link on the [Contributing to NumPy](https://numpy.org/devdocs/dev/index.html#stylistic-guidelines) page which should link directly to the NEP.

**Comment by bjnath** [comment]
I'm not sure where this leaves us. 

The original doc/C_STYLE_GUIDE.rst.txt is still in its original location, with new text that points the reader to the NEP:
```
The "NumPy C Style Guide" at this page has been supserseded by
"NEP 45 — C Style Guide" at https://numpy.org/neps/nep-0045-c_style_guide.html

```
1.  If  `doc/C_STYLE_GUIDE.rst.txt` was linked to before it should still be linked to. 
1. Is that even what we want?

**Comment by bjnath** [comment]
Sorry, I misunderstood -- you explained what you wanted.

> the Contributing to NumPy page...should link directly to the NEP.

**Comment by bjnath** [comment]
Opened PR #16500 to fix the link.
